File: packages/kdna-core/src/public-contract/digests.js
'use strict';

const { canonicalJson, parseJson, entryName, compareUtf8, utf8, reject } = require('./strict-input.js');

// Hashing is loaded only at the Core bytes boundary, never by Read's pure root.
function digest(bytes) {
  const { sha256 } = require('@noble/hashes/sha256');
  return 'sha256:' + Array.from(sha256(bytes), n => n.toString(16).padStart(2, '0')).join('');
}
function concat(parts) {
  const length = parts.reduce((n, p) => n + p.length, 0);
  if (!Number.isSafeInteger(length)) reject();
  const result = new Uint8Array(length); let offset = 0;
  for (const part of parts) { result.set(part, offset); offset += part.length; }
  return result;
}
function word(number, width) {
  if (!Number.isSafeInteger(number) || number < 0 || (width === 4 && number > 0xffffffff)) reject();
  const out = new Uint8Array(width), view = new DataView(out.buffer);
  if (width === 4) view.setUint32(0, number);
  else view.setBigUint64(0, BigInt(number));
  return out;
}
function admittedNames(entries) {
  const names = Object.keys(entries);
  if (names.some(n => !entryName(n) || n === 'build-receipt.json' || n.startsWith('reports/') || n.startsWith('authoring/'))) reject('READ_CORE_INVALID');
  return names.sort(compareUtf8);
}
function contentTreePreimage(entries) {
  const names = admittedNames(entries).filter(n => n !== 'checksums.json' && n !== 'signature.kdsig');
  const parts = [utf8('KDNA-CONTENT-TREE\0' + '0.2.0\0'), word(names.length, 4)];
  for (const name of names) {
    let content = entries[name]; const json = name.endsWith('.json');
    if (json) {
      const value = parseJson(content);
      if (name === 'kdna.json') {
        if (!value || typeof value !== 'object' || Array.isArray(value)) reject();
        delete value.content_digest;
        if (Object.hasOwn(value, 'authoring')) {
          if (!value.authoring || typeof value.authoring !== 'object' || Array.isArray(value.authoring)) reject();
          delete value.authoring.content_digest;
        }
      }
      content = utf8(canonicalJson(value));
    }
    const nameBytes = utf8(name);
    parts.push(word(nameBytes.length, 4), nameBytes, Uint8Array.of(json ? 0 : 1), word(content.length, 8), content);
  }
  return concat(parts);
}
function runtimeEntryNames(entries, manifest) {
  admittedNames(entries);
  const declared = manifest.runtime.mandatory_entries;
  if (new Set(declared).size !== declared.length || declared.some(n => !entryName(n) || ['checksums.json', 'signature.kdsig', 'mimetype'].includes(n))) reject('READ_CORE_INVALID');
  const names = [...new Set(['kdna.json', 'payload.kdnab', ...declared])].sort(compareUtf8);
  if (names.some(n => !Object.hasOwn(entries, n))) reject('READ_CORE_INVALID');
  return names;
}
function runtimeEntryPreimage(entries, manifest) {
  const names = runtimeEntryNames(entries, manifest);
  const parts = [utf8('KDNA-RUNTIME-ENTRY-SET\0' + '0.2.0\0'), word(names.length, 4)];
  for (const name of names) {
    const bytes = utf8(name), content = entries[name];
    parts.push(word(bytes.length, 4), bytes, word(content.length, 8), content);
  }
  return concat(parts);
}
function comparison(observed, expected = null, expected_source = null) {
  return expected === null ? { state: 'not_compared', expected: null, expected_source: null }
    : { state: observed === expected ? 'matched' : 'mismatched', expected, expected_source };
}
function evidence(key, observed, expected = null, expectedSource = null) {
  const basis = { A: 'container_bytes', C: 'content_tree', E: 'runtime_entry_set', P: 'runtime_capsule_jcs' }[key];
  const profile = key === 'P' ? 'kdna.canonicalization.runtime-capsule-jcs'
    : 'kdna.digest-basis.' + { A: 'container-bytes', C: 'content-tree', E: 'runtime-entry-set' }[key];
  return { basis, profile, profile_version: '0.2.0', algorithm: 'SHA-256', observed, comparison: comparison(observed, expected, expectedSource) };
}
function capsuleDigest(capsule) { return digest(utf8(canonicalJson(capsule))); }
module.exports = { digest, concat, word, contentTreePreimage, runtimeEntryPreimage, runtimeEntryNames, comparison, evidence, capsuleDigest };

File: python-sdk/kdna/core/_digests.py
import hashlib
from ._values import jcs, parse_json, entry_name, reject


def digest(data):
    return 'sha256:' + hashlib.sha256(data).hexdigest()


def word(n, width):
    if n < 0 or n > 9007199254740991 or (width == 4 and n > 0xffffffff):
        reject('READ_INPUT_INVALID')
    return n.to_bytes(width, 'big')


def content_tree(entries):
    names = sorted((n for n in entries if n not in ('checksums.json', 'signature.kdsig')), key=lambda x: x.encode())
    parts = [b'KDNA-CONTENT-TREE\x000.2.0\x00', word(len(names), 4)]
    for name in names:
        content = entries[name]
        is_json = name.endswith('.json')
        if is_json:
            value = parse_json(content)
            if name == 'kdna.json':
                value.pop('content_digest', None)
                if 'authoring' in value:
                    if type(value['authoring']) is not dict:
                        reject('READ_INPUT_INVALID')
                    value['authoring'].pop('content_digest', None)
            content = jcs(value).encode()
        n = name.encode()
        parts.extend([word(len(n), 4), n, bytes([0 if is_json else 1]), word(len(content), 8), content])
    return b''.join(parts)


def runtime_names(entries, manifest):
    declared = manifest['runtime']['mandatory_entries']
    if len(set(declared)) != len(declared) or any(not entry_name(n) or n in ('checksums.json', 'signature.kdsig', 'mimetype') for n in declared):
        reject()
    names = sorted(set(['kdna.json', 'payload.kdnab', *declared]), key=lambda x: x.encode())
    if any(n not in entries for n in names):
        reject()
    return names


def runtime_entries(entries, manifest):
    names = runtime_names(entries, manifest)
    parts = [b'KDNA-RUNTIME-ENTRY-SET\x000.2.0\x00', word(len(names), 4)]
    for name in names:
        n, content = name.encode(), entries[name]
        parts.extend([word(len(n), 4), n, word(len(content), 8), content])
    return b''.join(parts)


def evidence(key, observed, expected=None, expected_source=None):
    basis, suffix = {'A': ('container_bytes', 'container-bytes'), 'C': ('content_tree', 'content-tree'), 'E': ('runtime_entry_set', 'runtime-entry-set')}[key]
    comparison = {'state': 'not_compared', 'expected': None, 'expected_source': None}
    if expected is not None:
        comparison = {'state': 'matched' if expected == observed else 'mismatched', 'expected': expected, 'expected_source': expected_source}
    return {'basis': basis, 'profile': 'kdna.digest-basis.' + suffix, 'profile_version': '0.2.0', 'algorithm': 'SHA-256', 'observed': observed, 'comparison': comparison}
